sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/RoundRobinByAttributeRule.java

```diff
@@ -0,0 +1,197 @@
+package org.apache.mesos.offer.constrain;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.mesos.Protos.Attribute;
+import org.apache.mesos.Protos.Offer;
+import org.apache.mesos.Protos.TaskInfo;
+import org.apache.mesos.offer.AttributeStringUtils;
+import org.apache.mesos.offer.OfferRequirement;
+import org.apache.mesos.offer.TaskUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Implements logic for Marathon's GROUP_BY operator for attribute values. Ensures that tasks are
+ * evenly distributed across distinct attribute values in the system as they are rolled out.
+ *
+ * Example:
+ *  attribute |     tasks
+ * -----------+---------------
+ *    rack:1  | a-1, b-1, c-1
+ *    rack:2  | a-2, c-2, c-3
+ *    rack:3  | b-2, c-4
+ * Result:
+ *  allow rack:3 only, unless we know that there's >3 racks via the attribute_count parameter
+ *
+ * Example:
+ *  attribute |     tasks
+ * -----------+---------------
+ *    rack:1  | a-1, b-1, c-1
+ *    rack:2  | a-2, c-2, c-3
+ *    rack:3  | b-2, c-4, b-3
+ * Result:
+ *  allow any of rack:1/rack:2/rack:3, unless we know that there's >3 racks via the attribute_count
+ *  parameter.
+ */
+public class RoundRobinByAttributeRule implements PlacementRule {
```

Comment:
That case is meant to be covered by having the user specify the optional attributeValueCount in their filter. This allows us to know that there are other nodes out there which don't have anything on them at all yet. In case (known task nodes < user-provided expected node count) we avoid launching until we see an offer from a node which has that attribute with no tasks. See comment at L128 for handling that case.

If the user refrains from providing a number of nodes then we indeed will do that.